### PR TITLE
Initial docker-compose setup

### PR DESCRIPTION
**Pull request by sobering**
This is a work-in-progress, please do not merge. There are a few issues with this at the moment, I will attach comments to the relevant files/lines.

### CODE REVIEW

File: image/nginx/nginx.conf [hunk 1]
```diff
@@ -0,0 +1,100 @@
+user www-data;
```
Comment (sobering) [inline image/nginx/nginx.conf:1]: You previously had this set as root, noting a bug. Does that bug still exist? I had it working with `www-data` as the user.

File: image/nginx/nginx.conf [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+worker_processes 4;
+pid /run/nginx.pid;
```
Comment (sobering) [inline image/nginx/nginx.conf:3]: `pid` location is different in previous version. Still works?

File: image/php-fpm/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,24 @@
+FROM ubuntu
+
+# ADD /laravel /var/www
```
Comment (sobering) [inline image/php-fpm/Dockerfile:3]: You'll notice in `image/php-fpm/Dockerfile`, `image/cli/Dockerfile` and `image/nginx/Dockerfile`, I've commented out the `ADD` instruction. This is a bit tricky. We want all these images to share the `laravel` directory from the project root, but you can't add files outside the current build context. See https://github.com/docker/docker/issues/2745 for more info on this. @atrauzzi Your assistance with this would be appreciated.

File: image/php-fpm/php-fpm.conf [hunk 4]
```diff
@@ -86,7 +86,7 @@ error_log = /var/log/php5-fpm.log
 
 ; Send FPM to background. Set to 'no' to keep FPM in foreground for debugging.
 ; Default Value: yes
-daemonize = no
+;daemonize = yes
```
Comment (sobering) [inline image/php-fpm/php-fpm.conf]: This is an oversight on my part, I will change it back to `no`

File: image/php-fpm/php-fpm.conf [hunk 5]
```diff
@@ -22,7 +22,7 @@
 ; Pid file
 ; Note: the default prefix is /var
 ; Default Value: none
-pid = /run/php5-fpm.pid
+pid = /var/run/php5-fpm.pid
```
Comment (sobering) [inline image/php-fpm/php-fpm.conf:25]: Use previous value?

File: image/php-fpm/www.conf [hunk 6]
```diff
@@ -19,7 +19,7 @@
 ; Unix user/group of processes
 ; Note: The user is mandatory. If the group is not set, the default user's group
 ;       will be used.
-user = root
+user = www-data
```
Comment (sobering) [inline image/php-fpm/www.conf:22]: Use root? I believe I had this working with `www-data`.

File: image/php-fpm/www.conf [hunk 7]
```diff
@@ -41,8 +41,8 @@ listen = 9000
 ; BSD-derived systems allow connections regardless of permissions.
 ; Default Values: user and group are set as the running user
 ;                 mode is set to 0660
-;listen.owner = www-data
-;listen.group = www-data
+listen.owner = www-data
```
Comment (sobering) [inline image/php-fpm/www.conf:44]: Comment these out?

File: image/php-fpm/www.conf [hunk 8]
```diff
@@ -367,13 +367,11 @@ chdir = /
...
 
 env[APP_DEBUG] = $APP_DEBUG
-
-env[DB_HOST] = $POSTGRES_PORT_5432_TCP_ADDR
-env[POSTGRES_PORT] = $POSTGRES_PORT_5432_TCP_PORT
-env[DB_DATABASE] = $POSTGRES_DATABASE
-env[DB_USERNAME] = $POSTGRES_USER
-env[DB_PASSWORD] = $POSTGRES_PASSWORD
-
+env[DB_HOST] = $DB_PORT_5432_TCP_ADDR
```
Comment (sobering) [inline image/php-fpm/www.conf:370]: I gave an alias of `db` to the `postgres` container. That's why these env var names changed.

File: image/redis/Dockerfile [hunk 9]
```diff
@@ -0,0 +1 @@
+FROM redis
```
Comment (sobering) [inline image/redis/Dockerfile:1]: Forgot to add in our default config. I will add it back.

File: image/cli/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,28 @@
+FROM ubuntu
```
Comment (sobering) [inline image/cli/Dockerfile:1]: This file is incomplete. I need to add back the entrypoint, which sets the env vars. I also need to install node and some basic tools like `gulp`. @atrauzzi Question: why are we adding ssh keys to the cli container?

**Comment by sobering** [comment]
I should also make note that this PR does not include modifications for Darwin systems.

File: image/nginx/nginx.conf (hunk 1 above)
Comment (atrauzzi) [inline image/nginx/nginx.conf:1]: This would pertain to situations where you're running with a VM (OSX, Windows) and using the vboxsf filesystem driver to point back to your live code.

I think we spoke earlier about how to bind-mount directories for development mode.  Can you verify or just mention here whether this line is okay when you're bind-mounted (not just running off of files that were added to the container)?

File: image/nginx/nginx.conf (hunk 2 above)
Comment (atrauzzi) [inline image/nginx/nginx.conf:3]: Is that the debian-default?

File: image/php-fpm/Dockerfile (hunk 3 above)
Comment (atrauzzi) [inline image/php-fpm/Dockerfile:3]: In Docker 1.5 (newer than the issue you linked), support for multiple Dockerfiles per build context was added.  You and I spoke about this in the office, so I think we should be re-adding this line and  each container should ship with its own copy of code. :)

Comment (atrauzzi) [inline image/php-fpm/Dockerfile:3]: BTW, having an appreciation and understanding for this issue totally gives you +1 do your innate docker skills.

File: image/php-fpm/php-fpm.conf (hunk 4 above)
Comment (atrauzzi) [inline image/php-fpm/php-fpm.conf]: :)

File: image/php-fpm/php-fpm.conf (hunk 5 above)
Comment (atrauzzi) [inline image/php-fpm/php-fpm.conf:25]: Whatever the debian/ubuntu defaults isfor this.  I'm easy.  Just want to make sure we don't deviate from their standards so that we remain convention and tool compatible.

File: image/php-fpm/www.conf (hunk 6 above)
Comment (atrauzzi) [inline image/php-fpm/www.conf:22]: Same as previous.  Let's make sure this works when bind-mounted on OSX and/or Windows.

File: image/php-fpm/www.conf (hunk 8 above)
Comment (atrauzzi) [inline image/php-fpm/www.conf:370]: :+1:

File: image/cli/Dockerfile (hunk 10 above)
Comment (atrauzzi) [inline image/cli/Dockerfile:1]: Terrific question!

Head over here and read up on everything (yes, it's long).  I comment somewhere in the middle.  Then let's chat about it: https://github.com/docker/docker/issues/6396

tldr: We do this so that the cli container doesn't get stinkered by github (and others) rate limits for anonymous access.  Also allows containers to access private repositories.

File: image/php-fpm/Dockerfile (hunk 3 above)
Comment (atrauzzi) [inline image/php-fpm/Dockerfile:3]: See: https://github.com/docker/compose/issues/834

Comment (atrauzzi) [inline image/php-fpm/Dockerfile:3]: Also: https://github.com/atrauzzi/docker-laravel/blob/master/images/data/build#L17 anywhere it says `CONTEXT_ROOT`.

Comment (atrauzzi) [inline image/php-fpm/Dockerfile:3]: And finally: https://github.com/docker/compose/pull/1075

**Comment by sobering** [comment]
I'm thinking of revisiting this soon. Any thoughts/considerations you'd like me to be aware of before spending time on this again?

**Comment by atrauzzi** [comment]
Let's get up to speed with compose and whether it's moved at all over the last bit.

It might be worth understanding if there's anything that can be done to ensure mesos and kubernetes are satisfiable.
